execution/v0.1.0/execution.go

```diff
@@ -0,0 +1,175 @@
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+//    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+//    License for the specific language governing permissions and limitations
+//    under the License.
+
+package executionv010
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"golang.org/x/mod/semver"
+
+	common "github.com/cloudbase/garm-provider-common/execution/common"
+	"github.com/cloudbase/garm-provider-common/params"
+)
+
+func GetEnvironment() (EnvironmentV010, error) {
+	env := EnvironmentV010{
+		Command:            common.ExecutionCommand(os.Getenv("GARM_COMMAND")),
+		ControllerID:       os.Getenv("GARM_CONTROLLER_ID"),
+		PoolID:             os.Getenv("GARM_POOL_ID"),
+		ProviderConfigFile: os.Getenv("GARM_PROVIDER_CONFIG_FILE"),
+		InstanceID:         os.Getenv("GARM_INSTANCE_ID"),
+	}
+
+	// If this is a CreateInstance command, we need to get the bootstrap params
+	// from stdin
+	boostrapParams, err := common.GetBoostrapParamsFromStdin(env.Command)
+	if err != nil {
+		return EnvironmentV010{}, fmt.Errorf("failed to get bootstrap params: %w", err)
+	}
+	env.BootstrapParams = boostrapParams
+
+	if env.InterfaceVersion == "" {
+		env.InterfaceVersion = common.Version010
+	}
+
+	if err := env.Validate(); err != nil {
+		return EnvironmentV010{}, fmt.Errorf("failed to validate execution environment: %w", err)
+	}
+
+	return env, nil
+}
+
+type EnvironmentV010 struct {
+	Command            common.ExecutionCommand
+	ControllerID       string
+	PoolID             string
+	ProviderConfigFile string
+	InstanceID         string
+	InterfaceVersion   string
+	BootstrapParams    params.BootstrapInstance
+}
+
+func (e EnvironmentV010) Validate() error {
+	if e.Command == "" {
+		return fmt.Errorf("missing GARM_COMMAND")
+	}
+
+	if e.ProviderConfigFile == "" {
+		return fmt.Errorf("missing GARM_PROVIDER_CONFIG_FILE")
+	}
+
+	if _, err := os.Lstat(e.ProviderConfigFile); err != nil {
+		return fmt.Errorf("error accessing config file: %w", err)
+	}
+
+	if e.ControllerID == "" {
+		return fmt.Errorf("missing GARM_CONTROLLER_ID")
+	}
+
+	switch e.Command {
+	case common.CreateInstanceCommand:
+		if e.BootstrapParams.Name == "" {
+			return fmt.Errorf("missing bootstrap params")
+		}
+		if e.ControllerID == "" {
+			return fmt.Errorf("missing controller ID")
+		}
+		if e.PoolID == "" {
+			return fmt.Errorf("missing pool ID")
+		}
+	case common.DeleteInstanceCommand, common.GetInstanceCommand,
+		common.StartInstanceCommand, common.StopInstanceCommand:
+		if e.InstanceID == "" {
+			return fmt.Errorf("missing instance ID")
+		}
+	case common.ListInstancesCommand:
+		if e.PoolID == "" {
+			return fmt.Errorf("missing pool ID")
+		}
+	case common.RemoveAllInstancesCommand:
+		if e.ControllerID == "" {
+			return fmt.Errorf("missing controller ID")
+		}
+	case common.GetVersionCommand:
```

Comment:
The GetVersion command currently manually baked into the existing providers will return the version of the provider itself:

https://github.com/cloudbase/garm-provider-openstack/blob/ac46d4d5a542bca96cd0309c89437d3382c3ea26/main.go#L41-L49

Reply:
The interface version is already an exported field, which is parsed from the variables that the server sets, so there is little chance that the server will want to query that back. And a user will want to essentially get the version of the provider.